### PR TITLE
Use React.useId if available

### LINKED ISSUE #4607: Using <Suspense/> inside <SSRProvider/> causes hydration error
# 🐛 Bug Report 

Using `<Suspense/>` inside `<SSRProvider/>` cases this error:

```
Error: This Suspense boundary received an update before it finished hydrating. This caused the boundary to switch to client rendering. The usual way to fix this is to wrap the original update in startTransition.
```
AFAICT, this is caused by the `setState()` [in SSRProvider](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/ssr/src/SSRProvider.tsx#L71) in the layout effect which updates the value of the `SSRContext`; since `<Suspense/>` is a descendant and the context just changed, React has to re-render the content inside `<Suspense/>` as it can't rely on the server-generated html.

Here's a quick repro using Next.js: https://codesandbox.io/p/sandbox/affectionate-cache-01vp9h?file=%2Fpages%2Findex.tsx%3A8%2C20

## 🤔 Expected Behavior

No errors

## 😯 Current Behavior

```
Error: This Suspense boundary received an update before it finished hydrating. This caused the boundary to switch to client rendering. The usual way to fix this is to wrap the original update in startTransition.
```

## 💁 Possible Solution

Maybe use `startTransition` for the `setIsSsr()` call?  But not sure if that messes anything else up.

## 💻 Code Sample

https://codesandbox.io/p/sandbox/affectionate-cache-01vp9h?file=%2Fpages%2Findex.tsx%3A8%2C20

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-aria   | 3.25.0
| Next          | 13.4.1

## 🧢 Your Company/Team

Plasmic

### PR DESCRIPTION
**Pull request by devongovett**
Fixes #4607

This uses React 18's `useId` and `useSyncExternalStore` to replace our `useSSRSafeId` and `useIsSSR` hooks, with fallbacks for older reacts to our existing implementation. `SSRProvider` becomes a noop and apps can remove it if they are on React 18. This should improve support for streaming server rendering, as well as performance by removing the `useLayoutEffect` in the SSRProvider at the root of the app.

Two points in terms of backward compatibility:

1. This will change the format of everyone's ids, which will break snapshot tests. I don't think this can be avoided.
2. ~~Ids may start clashing if there are multiple versions of React on the page at once. For client only apps, we currently generate a random prefix to avoid this. With `React.useId`, this can be achieved using the [identifierPrefix](https://react.dev/reference/react/useId#specifying-a-shared-prefix-for-all-generated-ids) option of `React.createRoot`, but that's not something React Aria can control. So apps that upgrade which have multiple copies of React (e.g. micro frontends) may see ids clashing. This is something we need to verify.~~

To do

- [x] Update docs

### CODE REVIEW

**Comment by devongovett** [comment]
For point number 2, we could probably still prepend our own random prefix to the start of the id returned by `React.useId`. We would need to do this only when client rendering to avoid hydration mismatches, the same way we do today.

File: packages/@react-aria/ssr/src/SSRProvider.tsx [hunk 1]
```diff
@@ -79,6 +76,23 @@ export function SSRProvider(props: SSRProviderProps): JSX.Element {
...
 
+let warnedAboutSSRProvider = false;
+
+/**
+ * When using SSR with React Aria in React 16 or 17, applications must be wrapped in an SSRProvider.
+ * This ensures that auto generated ids are consistent between the client and server.
+ */
+export function SSRProvider(props: SSRProviderProps) {
+  if (typeof React['useId'] === 'function') {
+    if (process.env.NODE_ENV !== 'test' && !warnedAboutSSRProvider) {
```
Comment (snowystinger) [inline packages/@react-aria/ssr/src/SSRProvider.tsx:87]: How likely are other projects to test with SSR in their unit tests do you think? I guess it'd hopefully be the same between their tests and browser, so if they see the warnings in the test they'd see it in the browser anyways too

**Review by snowystinger — APPROVED** [review]
Checked some labels, table headers, and dialogs which use slot id's, all of it looks to be working well

File: packages/@react-aria/ssr/src/SSRProvider.tsx [hunk 2]
```diff
@@ -79,6 +76,23 @@ export function SSRProvider(props: SSRProviderProps): JSX.Element {
   );
 }
 
+let warnedAboutSSRProvider = false;
+
+/**
+ * When using SSR with React Aria in React 16 or 17, applications must be wrapped in an SSRProvider.
+ * This ensures that auto generated ids are consistent between the client and server.
+ */
+export function SSRProvider(props: SSRProviderProps) {
```
Comment (ktabors) [inline packages/@react-aria/ssr/src/SSRProvider.tsx:85]: Is there a reason you removed the return type here? `): JSX.Element {` 
`props.children` would also be a JSX.Element, right?

File: packages/react-aria-components/docs/react-aria-components.mdx [hunk 3]
```diff
@@ -323,8 +323,6 @@ import {Select, SelectValue, Label, Button, Popover, ListBox, Item} from 'react-
 
 </details>
 
-If you're using SSR via a framework like Next.js or Remix, you'll also need to wrap your application in an `<SSRProvider>`. See the [Server Side Rendering](ssr.html) docs for more details.
```
Comment (ktabors) [inline packages/react-aria-components/docs/react-aria-components.mdx:326]: Should we leave this but mention this is true when using react 16 or 17?

File: packages/@react-aria/ssr/src/SSRProvider.tsx (hunk 2 above)
Comment (devongovett) [reply packages/@react-aria/ssr/src/SSRProvider.tsx:85]: children can be a ReactNode, which could be a string, etc. in addition to an element. Anyway TS should auto-infer the type.

File: packages/react-aria-components/docs/react-aria-components.mdx (hunk 3 above)
Comment (devongovett) [reply packages/react-aria-components/docs/react-aria-components.mdx:326]: idk maybe. I figured most people using RAC will be using modern React, and if not they can find info on the SSR docs page. Putting lots of caveats all over the place might be more confusing. Not sure.

File: packages/@react-stately/datepicker/src/utils.ts [hunk 4]
```diff
@@ -143,8 +138,13 @@ export function useDefaultProps(v: DateValue, granularity: Granularity): [Granul
...
 
-  // If the granularity or time zone changed, update the last value.
-  if (!lastValue || lastValue[0] !== granularity || lastValue[1] !== defaultTimeZone) {
+  let [lastValue, setLastValue] = useState<[Granularity, string]>([granularity, defaultTimeZone]);
+  if (!v) {
+    return lastValue;
+  }
+
+    // If the granularity or time zone changed, update the last value.
+  if (lastValue[0] !== granularity || lastValue[1] !== defaultTimeZone) {
```
Comment (LFDanLu) [inline packages/@react-stately/datepicker/src/utils.ts:147]: Looks like the changes here have made it so changes to a DateField's granularity doesn't get reflected after the initial render: https://reactspectrum.blob.core.windows.net/reactspectrum/c66f9b0844320f6a8166d74881b5ec629aee0c5c/storybook/index.html?path=/story/date-and-time-datefield--default&args=granularity:hour&providerSwitcher-express=false

**Review by LFDanLu — CHANGES_REQUESTED** [review]

**Review by LFDanLu — APPROVED** [review]

**Review by snowystinger — APPROVED** [review]

**Comment by ssijak** [comment]
@devongovett Thanks for this. How can I track/know when this change will be released? Need it because of next13 `/app` dir

**Comment by devongovett** [comment]
It'll be in our next stable release in the next few weeks. In the meantime, you can try out the nightly version if you like.

**Comment by dkryaklin** [comment]
@devongovett added PR that addresses this problem https://github.com/adobe/react-spectrum/pull/4738

**Pull request merged**